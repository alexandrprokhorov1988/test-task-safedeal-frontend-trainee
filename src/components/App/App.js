import React from 'react';
import './App.css';
import Header from '../../components/Header/Header';
import PhotoGrid from '../../components/PhotoGrid/PhotoGrid';
import Footer from '../../components/Footer/Footer';
import Popup from '../../components/Popup/Popup';
import Preloader from '../../components/Preloader/Preloader';
import mainApi from "../../utils/mainApi";
import {dateParseFromTimestampToString} from "../../utils/helpers";

function App() {
  const [cards, setCards] = React.useState([]);
  const [isLoading, setIsLoading] = React.useState(false);
  const [isLoadingComment, setIsLoadingComment] = React.useState(false);
  const [isOpenPopup, setIsOpenPopup] = React.useState(false);
  const [currentOriginSizeImage, setCurrentOriginSizeImage] = React.useState({});
  const [comments, setComments] = React.useState([]);

  React.useEffect(() => {
    setIsLoading(true);
    mainApi.getInitialImages()
      .then((data) => {
        setCards(data);
      })
      .catch((err) => {
        console.log(err);
      })
      .finally(() =>
        setIsLoading(false)
      )
  }, []);

  function handlePopupOpen() {
    setIsOpenPopup(true);
    document.addEventListener('keydown', handleEscClose);
    document.body.style.overflow = "hidden";
  }

  function closePopup() {
    setIsOpenPopup(false);
    document.removeEventListener('keydown', handleEscClose);
    document.body.style.overflow = "auto";
  }

  function handleEscClose(e) {
    if (e.key === 'Escape') {
      closePopup();
    }
  }

  function handleGetOriginalSizeImage(imageId) {
    if (imageId === currentOriginSizeImage.id) {
      handlePopupOpen();
    } else {
      setCurrentOriginSizeImage({});
      return mainApi.getOriginalSizeImage(imageId)
        .then((res) => {
          setCurrentOriginSizeImage({
            id: res.id,
            url: res.url
          });
          const newComments = res.comments.map((comment) => ({
            text: comment.text,
            date: dateParseFromTimestampToString(comment.date),
            id: comment.id
          }));
          setComments(newComments);
          handlePopupOpen();
        })
        .catch((err) => console.log(err));
    }
    return true;
  }

  function handleSubmitComment(values, imageId) {
    setIsLoadingComment(true);
    return mainApi.setNewComment({
      id: imageId,
      name: values.name,
      comment: values.comment
    })
      .then(() => {
        const answerFromServer = {
          id: Math.random(),
          text: values.comment,
          date: dateParseFromTimestampToString(new Date().getTime())
        };
        setComments([...comments, answerFromServer]);
      })
      .catch((err) => console.log(err))
      .finally(() => setIsLoadingComment(false));
  }

  return (
    <div className="page">
      <Header/>
      <main className="content">
        {isLoading ? <Preloader/> :
          <PhotoGrid
            cards={cards}
            isLoading={isLoading}
            onImageClick={handleGetOriginalSizeImage}
          />
        }
      </main>
      <Footer/>
      <Popup
        isOpenPopup={isOpenPopup}
        onClose={closePopup}
        isLoadingComment={isLoadingComment}
        onSubmitComment={handleSubmitComment}
        comments={comments}
        currentOriginSizeImage={currentOriginSizeImage}
      />
    </div>
  );
}

export default App;
