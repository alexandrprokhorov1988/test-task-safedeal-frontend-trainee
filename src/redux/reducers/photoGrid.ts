import {
  photoGridActions,
  SET_CARDS,
  SET_COMMENTS,
  SET_CURRENT_ORIGIN_SIZE_IMAGE,
  SET_IS_LOADING
} from "../actions/photoGridActions";

interface IPhotoGridState {
  cards: any[];
  isLoading: boolean;
  currentOriginSizeImage: {
    id?: number,
    url?: string
  };
  comments: any[];
}

export const initialState: IPhotoGridState = {
  cards: [],
  isLoading: false,
  currentOriginSizeImage: {},
  comments: [],
};

export function photoGridReducer(state = initialState, action: photoGridActions): IPhotoGridState {
  switch (action.type) {
    case SET_CARDS:
      return {...state, cards: action.payload};
    case SET_CURRENT_ORIGIN_SIZE_IMAGE:
      return {...state, currentOriginSizeImage: action.payload};
    case SET_COMMENTS:
      return {...state, comments: action.payload};
    case SET_IS_LOADING:
      return {...state, isLoading: action.payload};
    default:
      return state;
  }
}
