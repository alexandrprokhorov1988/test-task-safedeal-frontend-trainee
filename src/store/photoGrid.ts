import {makeAutoObservable} from "mobx";
import {autoSave} from "../utils/helpers";

class PhotoGrid {
  public cards: any[] = [];

  public isLoading = false;

  public originSizeImage: { id?: number, url?: string } = {};

  public comments: { id: number, text: string, date: string }[] = [];

  constructor() {
    makeAutoObservable(this);
    autoSave(this, 'photoGridStore')
  }

  public setCards(newCard: any[]) {
    this.cards = newCard;
  }

  public setIsLoading(state: boolean) {
    this.isLoading = state;
  }

  public setOriginSizeImage(img: { id?: number, url?: string }) {
    this.originSizeImage = img;
  }

  public setComments(comments: { id: number, text: string, date: string }[]) {
    this.comments = comments;
  }
}

export default new PhotoGrid();
