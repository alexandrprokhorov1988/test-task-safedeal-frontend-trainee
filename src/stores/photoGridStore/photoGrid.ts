import {makeAutoObservable} from "mobx";
import {autoSave} from "../../utils/helpers";
import photoGridService from '../photoGridStore/photoGrid.service';
import {NOT_FOUND_ERR} from "../../utils/constants";

class PhotoGrid {
  public cards: any[] = [];

  public isLoading = false;

  constructor() {
    makeAutoObservable(this);
    autoSave(this, 'photoGridStore');
  }

  public setIsLoading(state: boolean) {
    this.isLoading = state;
  }

  public setCards(arr: any[]) {
    this.cards = arr;
  }

  public getInitialImages() {
    this.isLoading = true;
    return photoGridService.getInitialImages()
      .then((response) => {
        this.cards = response.data;
      })
      .catch((err) => {
        if (err.response.status === 404) {
          console.log(NOT_FOUND_ERR)
        } else {
          console.log(err.message);
        }
      })
      .finally(() => {
        this.isLoading = false;
      })
  }
}

export default new PhotoGrid();
