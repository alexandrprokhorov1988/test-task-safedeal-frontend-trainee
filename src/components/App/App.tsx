import React from 'react';
import './App.css';
import Header from '../../components/Header/Header';
import Footer from '../../components/Footer/Footer';
import PhotoGrid from '../../components/PhotoGrid/PhotoGrid';
import Popup from "../Popup/Popup";
import app from '../../store/app';
import {dateParseFromTimestampToString} from "../../utils/helpers";

const App = () => {
console.log(dateParseFromTimestampToString(1621278139402));
  function handlePopupOpen(): void {
    app.setIsOpenPopup(true);
    document.addEventListener('keydown', handleEscClose);
    document.body.style.overflow = "hidden";
  }

  function closePopup(): void {
    app.setIsOpenPopup(false);
    document.removeEventListener('keydown', handleEscClose);
    document.body.style.overflow = "auto";
  }

  function handleEscClose(event: any): void {
    if (event.key === 'Escape') {
      closePopup();
    }
  }

  return (
    <div className="page">
      <Header/>
      <main className="content">
        <PhotoGrid onOpen={handlePopupOpen}/>
      </main>
      <Footer/>
      <Popup onClose={closePopup}/>
    </div>
  );
};

export default App;
