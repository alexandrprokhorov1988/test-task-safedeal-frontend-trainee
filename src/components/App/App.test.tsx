import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import photoGrid from "../../stores/photoGridStore";

describe('App component', () => {
  photoGrid.setCards(
    [
      {id: 1, url: '#/test1'},
      {id: 5, url: '#/test5'}
    ]
  );
  it('App snapshot', () => {
    const render = renderer
      .create(<App/>)
      .toJSON();
    expect(render).toMatchSnapshot();
  });
});
