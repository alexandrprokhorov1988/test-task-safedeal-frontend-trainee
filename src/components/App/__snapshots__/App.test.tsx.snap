// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App snapshot 1`] = `
<div
  className="page"
>
  <footer
    className="header"
  >
    <h1
      className="header__title"
    >
      Test APP
    </h1>
  </footer>
  <main
    className="content"
  >
    <section
      className="photo-grid"
    >
      <img
        alt="Картинка"
        className="photo-grid-img"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="presentation"
        src="#/test1"
        tabIndex={0}
      />
      <img
        alt="Картинка"
        className="photo-grid-img"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="presentation"
        src="#/test5"
        tabIndex={0}
      />
    </section>
  </main>
  <footer
    className="footer"
  >
    <p
      className="footer__text"
    >
      © 2018-2019-2021
    </p>
  </footer>
  <div
    aria-hidden="true"
    className="popup "
    onClick={[Function]}
    role="button"
    tabIndex={-1}
  >
    <div
      aria-hidden="true"
      className="popup__container"
      onClick={[Function]}
      role="complementary"
      tabIndex={-1}
    >
      <button
        className="popup__close-button"
        onClick={[Function]}
        type="button"
      />
      <img
        alt="Картинка"
        className="popup__img"
      />
      <div
        className="popup__comments-container"
      />
      <form
        action="#"
        className="popup__form"
        method="get"
        noValidate={true}
        onSubmit={[Function]}
      >
        <span
          className="popup__form-error popup__form-error_hide"
        >
          
        </span>
        <input
          className="popup__form-input"
          maxLength={200}
          minLength={2}
          name="name"
          onChange={[Function]}
          pattern="^[а-яёА-ЯЁa-zA-Z0-9-\\\\s\\\\-]+$"
          placeholder="Ваше имя"
          required={true}
          type="text"
          value=""
        />
        <span
          className="popup__form-error popup__form-error_hide"
        >
          
        </span>
        <input
          className="popup__form-input"
          maxLength={200}
          minLength={2}
          name="comment"
          onChange={[Function]}
          pattern="^[а-яёА-ЯЁa-zA-Z0-9-\\\\s\\\\-]+$"
          placeholder="Ваш комментарий"
          required={true}
          type="text"
          value=""
        />
        <button
          className="popup__form-button"
          disabled={true}
          type="submit"
        >
          Оставить комментарий
        </button>
      </form>
    </div>
  </div>
</div>
`;
